classes/cliTool/ConvertLogFile.php

```diff
@@ -0,0 +1,1012 @@
+        }
+    }
+
+    /**
+     * Get the folder the log file is in.
+     */
+    abstract public function getLogFileDir(): string;
+
+    /**
+     * Get regular expression to parse log file entries
+     */
+    abstract public function getParseRegex(): string;
+
+    /**
+     * Get the datetime format used in the log file
+     */
+    abstract public function getPhpDateTimeFormat(): string;
+
+    /**
+     * Weather the URL parameters are used instead of CGI PATH_INFO.
+     * This will determine how URLs are parsed.
+     */
+    abstract public function isPathInfoDisabled(): bool;
+
+    /**
+     * Weather this is an apache access log file
+     */
+    abstract public function isApacheAccessLogFile(): bool;
+
+    /**
+     * Convert log file into the new format.
+     *
+     * The old log file will be renamed: '_old' is added at the end of the file name.
+     */
+    public function convert(string $fileName): void
+    {
+        $filePath = $this->getLogFileDir() . '/' . $fileName;
+
+        $path_parts = pathinfo($filePath);
+        $extension = $path_parts['extension'];
+
+        $newFilePath = $this->getLogFileDir() . '/' . $path_parts['filename'] . '_new.log';
+        if ($extension == 'gz') {
+            $fileMgr = new FileManager();
+            try {
+                $filePath = $fileMgr->gzDecompressFile($filePath);
+            } catch (Exception $e) {
+                printf($e->getMessage() . "\n");
+                exit(1);
+            }
+        }
+
+        $fhandle = fopen($filePath, 'r');
+        if (!$fhandle) {
+            echo "Error: Can not open file {$filePath}.\n";
+            exit(1);
+        }
+        $lineNumber = 0;
+        $isSuccessful = false;
+        while (!feof($fhandle)) {
+            $newEntry = [];
+            $lineNumber++;
+            $line = trim(fgets($fhandle));
+            if (empty($line) || substr($line, 0, 1) === '#') {
+                continue;
+            } // Spacing or comment lines.
+
+            $entryData = $this->getDataFromLogEntry($line);
+
+            if (!$this->isLogEntryValid($entryData)) {
+                echo "Invalid log entry at line {$lineNumber}.\n";
+                continue;
+            }
+
+            // Avoid internal apache requests.
+            if ($entryData['url'] == '*') {
+                continue;
+            }
+
+            // Avoid non sucessful requests.
+            $sucessfulReturnCodes = [200, 304];
+            if (!in_array($entryData['returnCode'], $sucessfulReturnCodes)) {
+                continue;
+            }
+
+            $newEntry['time'] = $entryData['date'];
+
+            $ip = $entryData['ip'];
+            $ipNotHashed = preg_match('/^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/', $ip);
+            // shell IPv6 be considered ?
+            if ($ipNotHashed === 1) {
+                $saltFileName = StatisticsHelper::getSaltFileName();
+                $salt = trim(file_get_contents($saltFileName));
```

Comment:
I change it so that it is only once read for a file, and not for each line.
Then later I will ask you something about that salt file generally -- maybe it can be only in the cache...